ezmlm-clean.1
.TH ezmlm-clean 1
.SH NAME
ezmlm-clean \- clean moderation directory
.SH SYNOPSIS
.B ezmlm-clean [-mMrRvV]
.I dir
.SH DESCRIPTION
.B ezmlm-clean
sends out messages for timed-out posts in the moderation directory,
removes timed-out messages awaiting confirm,
and removes stubs for rejected and accepted posts
for the mailing list stored in
.IR dir .

.B ezmlm-clean
reads
.I dir\fB/modtime
and extracts a time-out ``time'' in hours from it. If ``time'' is 0 or
.I dir\fB/modtime
is empty or doesn't exist,
a default of 120 h is used. If a time is given, it is limited to
the range 24 h to 240 h.

.B ezmlm-clean
then looks through
.IR dir\fB/mod/accepted/ ,
.IR dir\fB/mod/rejected/ ,
.IR dir\fB/mod/unconfirmed/ ,
and
.I dir\fB/mod/pending/
and removes message stubs older than ``time''. ``Time'' is
a minimum retention time. Since the files are processed only
when
.B ezmlm-clean
is run, the delay before a message is timed-out may
be substantially longer if the list does not receive many messages.

Message age
determined by the time parsed
.I from the file name,
not from the creation time.
Thus, there is no good way to extend the life of
the file by e.g. touching it. Also, files in these directories are not
checked for the proper format. Thus, most non-message files in these
directories will be deleted the first time
.B ezmlm-clean
is run.

For messages in
.I dir\fB/mod/pending/
and
.IR dir\fB/mod/unconfirmed/ ,
no action is taken on read-only files. Messages without the owner execute
bit set are silently removed, as they are the result of incomplete
.B ezmlm-store(1)
executions.
For other messages in
.IR dir\fB/mod/pending/ ,
a notification of the time out is sent
to the sender, before the file is removed.
Messages in
.IR dir\fB/mod/unconfirmed/ ,
.IR dir\fB/mod/accepted/ ,
and
.I dir\fB/mod/rejected/
are silently removed.

.B ezmlm-clean
logs errors to the mail log.
Re-delivery should be avoided by suffixing any
.I \.qmail
line invoking
.B ezmlm-clean
with '|| exit 0'.

By default, the timed-out post is returned to the sender.  If
.I dir\fB/noreturnposts
exists, it switches the default to discard the timed-out post.
.B Note:
.B ezmlm-clean
is normally run from both
.I dir\fB/editor
and
.IR dir\fB/moderator .
To suppress sender notification, either
.I dir\fB/noreturnposts
needs to exist or the
.B \-R
option needs to be specified for all invocations of
.BR ezmlm-clean .
.SH OPTIONS
.TP
.B \-m
(Default.)
The timed-out post is sent as a MIME enclosure.
.TP
.B \-M
The timed-out post is appended to the message.
.TP
.B \-r
(Default.)
The timed-out post is returned to sender.
.TP
.B \-R
The timed-out post is discarded without sender notification.
.TP
.B \-v
Display
.B ezmlm-clean
version information.
.TP
.B \-V
Display
.B ezmlm-clean
version information.
.SH "CHARACTER SETS"
If
.I dir\fB/charset
exists,
.B ezmlm-clean
will use the character set listed for all messages. Otherwise, the
default ``us-ascii'' will be used. The character set can be suffixed
by ``:'' followed by a code. If the code is ``Q'', outgoing messages are 
sent as ``Quoted-Printable'', if it is ``B'' they are sent ``base64'' encoded.
Otherwise, text is sent as is.
.SH "SEE ALSO"
ezmlm-confirm(1),
ezmlm-make(1),
ezmlm-moderate(1),
ezmlm-store(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================


ezmlm-clean.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "error.h"
#include "stralloc.h"
#include "str.h"
#include "env.h"
#include "sig.h"
#include "getconf.h"
#include "strerr.h"
#include "byte.h"
#include "getln.h"
#include "case.h"
#include "qmail.h"
#include "substdio.h"
#include "readwrite.h"
#include "seek.h"
#include "quote.h"
#include "datetime.h"
#include "now.h"
#include "direntry.h"
#include "cookie.h"
#include "sgetopt.h"
#include "fmt.h"
#include "messages.h"
#include "copy.h"
#include "open.h"
#include "scan.h"
#include "lock.h"
#include "hdr.h"
#include "die.h"
#include "idx.h"
#include "mime.h"
#include "config.h"
#include "auto_version.h"

int flagmime = MOD_MIME;	/* default is message as attachment */
int flagreturn = -1;
stralloc fnmsg = {0};

/* When ezmlm-clean is run, messages and message stubs in pending/      */
/* rejected/accepted/unconfirmed are erased if they are older than      */
/* delay hours.                                                         */
/* Timeouts in h for messages. If modtime has a number, it is made to be*/
/* in the range DELAY_MIN..DELAY_MAX. If the number is 0 or there is no */
/* number, DELAY_DEFAULT is used. Messages that are read-only are       */
/* ignored. Messages in 'pending' that have the execute bit set result  */
/* in an informative reply to the poster. Any defects in the message    */
/* format, inability to open the file, etc, result in a maillog entry   */
/* whereafter the message is erased. */

/* The defines are in "idx.h" */

const char FATAL[] = "ezmlm-clean: fatal: ";
const char USAGE[] =
"ezmlm-clean: usage: ezmlm-clean [-mMrRvV] dir";

void die_read(void)
{
  strerr_die2x(111,FATAL,MSG1(ERR_READ,fnmsg.s));
}

datetime_sec when;
unsigned int older;

char textbuf[1024];
substdio sstext;

struct qmail qq;

char *dir;
char strnum[FMT_ULONG];
char boundary[COOKIE];
datetime_sec hashdate;

stralloc quoted = {0};
stralloc line = {0};
stralloc modtime = {0};
stralloc to = {0};

int fd;
int match;
unsigned long msgnum = 0;
			/* counter to make message-id unique, since we may */
			/* send out several msgs. This is not bullet-proof.*/
			/* Duplication occurs if we do x>1 msg && another  */
			/* ezmlm started within x seconds, and with the    */
			/* same pid. Very unlikely.                        */

void sendnotice(const char *d)
/* sends file pointed to by d to the address in the return-path of the  */
/* message. */
{
  unsigned int x,y;
  const char *err;

      if (qmail_open(&qq, (stralloc *) 0) == -1)
        strerr_die2sys(111,FATAL,MSG(ERR_QMAIL_QUEUE));

      fd = open_read(d);
      if (fd == -1)
        strerr_die2sys(111,FATAL,MSG1(ERR_OPEN,d));
      substdio_fdbuf(&sstext,read,fd,textbuf,sizeof(textbuf));
      if (getln(&sstext,&line,&match,'\n') == -1) die_read();
      if (!match) die_read();
      if (!case_startb(line.s,line.len,"return-path:")) die_read();
      x = 12 + byte_chr(line.s + 12,line.len-12,'<');
      y = byte_rchr(line.s + x,line.len-x,'>');
      if (x != line.len && x+y != line.len) {
        if (!stralloc_copyb(&to,line.s+x+1, y-1)) die_nomem();
        if (!stralloc_0(&to)) die_nomem();
      } else
        die_read();
      hdr_add2s("Mailing-List: ",MSG(TXT_MAILING_LIST));
      if (listid.len > 0)
	hdr_add2("List-ID: ",listid.s,listid.len);
      hdr_datemsgid(when+msgnum++);
      hdr_from("-help");
      hdr_subject(MSG(SUB_RETURNED_POST));
      hdr_add2s("To: ",to.s);
      if (flagmime) {
	hdr_mime(CTYPE_MULTIPART);
	hdr_boundary(0);
	hdr_ctype(CTYPE_TEXT);
        hdr_transferenc();
      } else
      qmail_puts(&qq,"\n\n");

      copy(&qq,"text/top",flagcd);
      copy(&qq,"text/mod-timeout",flagcd);
      if (flagcd == 'B') {
        encodeB("",0,&line,2);
        qmail_put(&qq,line.s,line.len);
      }

      if (flagmime) {
	hdr_boundary(0);
	hdr_ctype(CTYPE_MESSAGE);
        qmail_puts(&qq,"\n");
      }

      if (seek_begin(fd) == -1)
        strerr_die2sys(111,FATAL,MSG1(ERR_SEEK,d));

      substdio_fdbuf(&sstext,read,fd,textbuf,sizeof(textbuf));
      if (qmail_copy(&qq,&sstext,-1) != 0) die_read();
      close (fd);

      if (flagmime)
	hdr_boundary(1);

      if (!stralloc_copy(&line,&outlocal)) die_nomem();
      if (!stralloc_cats(&line,"-return-@")) die_nomem();
      if (!stralloc_cat(&line,&outhost)) die_nomem();
      if (!stralloc_0(&line)) die_nomem();
      qmail_from(&qq,line.s);		/* sender */
        qmail_to(&qq,to.s);

     if (*(err = qmail_close(&qq)) != '\0')
       strerr_die4x(111,FATAL,MSG(ERR_TMP_QMAIL_QUEUE),": ", err + 1);

     strnum[fmt_ulong(strnum,qmail_qp(&qq))] = 0;
     strerr_warn2("ezmlm-clean: info: qp ",strnum,0);
}

void dodir(const char *dirname,int reply)
/* parses file names in directory 'dirname'. Files that are not owner */
/* writable (w) are ignored. If the files are older (by name!) than   */
/* now-delay, action is taken:                                        */
/* If the owner x bit is not set, the file is erased.                 */
/* If it is set and reply is not set, the file is erased. If both are */
/* set, a notice about the timeout is sent to the poster. If this     */
/* fails due to a message-related error (format, etc) the file is     */
/* erased even though no notice is sent. For temporary errors (like   */
/* out-of-memory) the message is left intact for the next run. If the */
/* notice is sent successfully, the file is erased. All this is to    */
/* do the best possible without risking a rerun of the .qmail file,   */
/* which could result in a redelivery of the action request and a     */
/* second (incorrect) reply to the moderator's request.               */

/* NOTE: ALL non-hidden files in this dir are processed and merci-    */
/* lessly deleted. No checks for proper file name. E.g. 'HELLO'       */
/* => time 0 => will be deleted on the next ezmlm-clean run.          */
{
  DIR *moddir;
  direntry *d;
  unsigned long modtime;
  struct stat st;

  moddir = opendir(dirname);
  if (!moddir)
    strerr_die2sys(0,FATAL,MSG1(ERR_OPEN,dirname));
  while ((d = readdir(moddir))) {
    if (d->d_name[0] == '.') continue;
    scan_ulong(d->d_name,&modtime);
    if (modtime < older) {
      if (!stralloc_copys(&fnmsg,dirname)) die_nomem();
      if (!stralloc_cats(&fnmsg,d->d_name)) die_nomem();
      if (!stralloc_0(&fnmsg)) die_nomem();
      if((stat(fnmsg.s,&st) != -1) && (st.st_mode & 0200)) {
        if(reply && (st.st_mode & 0100)) {
			/* unlink unless there was a TEMPORARY */
			/* not message-related error notifying */
			/* poster and msg x bit set.  Leave r/o*/
			/* messages alone. Non-x bit msg are   */
			/* trash. Just unlink, don't notify    */
          sendnotice(fnmsg.s);
          unlink(fnmsg.s);
        } else
          unlink(fnmsg.s);
      }
    }
  }
  closedir(moddir);
}


void main(int argc,char **argv)
{
  int fdlock;
  unsigned long delay;
  int opt;
  (void) umask(022);
  sig_pipeignore();
  when = now();

  while ((opt = getopt(argc,argv,"mMrRvV")) != opteof)
    switch(opt) {
      case 'm': flagmime = 1; break;
      case 'M': flagmime = 0; break;
      case 'r': flagreturn = 1; break;
      case 'R': flagreturn = 0; break;
      case 'v':
      case 'V': strerr_die2x(0,"ezmlm-clean version: ", auto_version);
                /* not reached */
      default:
	die_usage();
    }

  startup(dir = argv[optind]);
  if (flagreturn < 0)
    /* default to returning timed-out messages */
    flagreturn = !getconf_isset("noreturnposts");

  getconf_line(&modtime,"modtime",0);
  if (!stralloc_0(&modtime)) die_nomem();
  scan_ulong(modtime.s,&delay);
  if (!delay) delay = DELAY_DEFAULT;
  else if (delay < DELAY_MIN) delay = DELAY_MIN;
  else if (delay > DELAY_MAX) delay = DELAY_MAX;
  older = (unsigned long) when - 3600L * delay;	/* delay is in hours */

  fdlock = lockfile("mod/lock");

  dodir("mod/pending/",flagreturn);
  dodir("mod/accepted/",0);
  dodir("mod/rejected/",0);
  dodir("mod/unconfirmed/",0);
  _exit(0);
}

